rho/facts.py

```diff
@@ -0,0 +1,242 @@
+# Copyright (c) 2017 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+
+# At one point we had a bug where we added a new fact but forgot to
+# add it to ALL_FACTS, so users couldn't select it. The way we define
+# facts below is intended to make it impossible for that to happen
+# again.
+
+"""Defining the set of facts that Rho can scan for."""
+
+import collections
+import sys
+from rho.utilities import log
+
+ALWAYS_COLLECT = set()
+SENSITIVE_FACTS = set()
+DEFAULT_FACTS = set()
+ALL_FACTS = set()
+
+RHEL_FACTS = set()
+JBOSS_FACTS = set()
+
+# 'rho fact list' will print fact documentation in whatever order this
+# dictionary's iterator returns them, so need to keep it sorted.
+FACT_DOCS = collections.OrderedDict()
+
+
+def expand_facts(fact_names):
+    """Expand a list of facts with implicit facts and short names.
+
+    :param fact_names: a list of fact names
+    :returns: a set of facts to collect
+    """
+
+    facts_to_collect = set()
+
+    if fact_names == []:
+        facts_to_collect.update(DEFAULT_FACTS)
+
+    for fact in fact_names:
+        if fact == 'all':
+            return ALL_FACTS
+
+        if fact == 'default':
+            facts_to_collect.update(DEFAULT_FACTS)
+        elif fact == 'rhel':
+            facts_to_collect.update(RHEL_FACTS)
+        elif fact == 'jboss':
+            facts_to_collect.update(JBOSS_FACTS)
+        else:
+            facts_to_collect.add(fact)
+
+    facts_to_collect.update(ALWAYS_COLLECT)
+
+    if not facts_to_collect.issubset(ALL_FACTS):
+        log.error('Bad fact list %s', facts_to_collect)
```

Comment:
It would be nicer here if we output the difference that shows the facts that are actually invalid in the subset.

Reply:
Done, thanks! Sample output:
```
> PYTHONPATH=rho:$PYTHONPATH bin/rho scan --profile=jboss-test-server --facts jboss.eap.installed-versions --facts jboss.eap.deploy-dates --reportfile=out.csv -vvv --facts fizzbuzz --facts abc
Unknown facts requested: abc, fizzbuzz
```